azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -44,3 +58,327 @@ def _execute_op(self, op):
+        registration_state = None
+        if decoded_state is not None:
+            # Everything needs to be converted to string explicitly for python 2
+            # as everything is by default a unicode character
+
+            registration_state = RegistrationState(
+                device_id=get_optional_element(decoded_state[0], "deviceId"),
+                assigned_hub=get_optional_element(decoded_state[0], "assignedHub"),
+                sub_status=get_optional_element(decoded_state[0], "substatus"),
+                created_date_time=get_optional_element(decoded_state[0], "createdDateTimeUtc"),
+                last_update_date_time=get_optional_element(
+                    decoded_state[0], "lastUpdatedDateTimeUtc"
+                ),
+                etag=get_optional_element(decoded_state[0], "etag"),
+                payload=get_optional_element(decoded_state[0], "payload"),
+            )
+
+        registration_result = RegistrationResult(
+            operation_id=operation_id, status=status, registration_state=registration_state
+        )
+        return registration_result
+
+
+class RegistrationStage(CommonProvisioningStage):
+    """
+    This is the first stage that decides converts a registration request
+    into a normal request and response operation.
+    Upon the receipt of the response this stage decides whether
+    to send another registration request or send a query request.
+    Depending on the status and result of the response
+    this stage may also complete the registration process.
+    """
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _execute_op(self, operation):
+        if isinstance(operation, pipeline_ops_provisioning.SendRegistrationRequestOperation):
+            initial_register_op = operation
+
+            def on_registration_response(op, error):
+                logger.debug(
+                    "{stage_name}({op_name}): Received response with status code {status_code} for SendRegistrationRequestOperation".format(
+                        stage_name=self.name, op_name=op.name, status_code=op.status_code
+                    )
+                )
+
+                # This could be an error that has been reported to this stage Or this
+                # could be an error because the service responded with status code >=300
+                error = self._get_error(op, REGISTER, error=error)
+
+                if not error:
+                    success_status_code = op.status_code
+                    retry_interval = (
+                        int(op.retry_after, 10)
+                        if op.retry_after is not None
+                        else constant.DEFAULT_POLLING_INTERVAL
+                    )
+                    decoded_response = self._decode_response(op)
+                    operation_id = self._get_operation_id(decoded_response)
+
+                    # retry after scenario
+                    if success_status_code >= 429:
```

Comment:
>if success_status_code >= 429: [](start = 20, length = 30)

This confuses me.  You're treating everything 429+ as success, but then treating the 429+ failures as a special kind of success that's not really a success.

Reply:
may be we need to speak. This was the exact behavior in polling state machine. Anything >= 429 was going to retry the process again in the PUT Register as well as GET Query. It is not a true error unless I am mistaken

---
In reply to: [350962570](https://github.com/Azure/azure-iot-sdk-python/pull/391#discussion_r350962570) [](ancestors = 350962570)